### DISCUSSION TITLE
batchfd?

### DISCUSSION THREAD

**Discussion by stsp** [c-blake/batch#3] [Ideas]
Hi!

I came across this project while searching
for something completely different. :)
Namely, I want to assign some syscalls to
be executed when program abnormally
terminates (even killed by SIGKILL).
So I was looking for some batchfd() syscall,
which would allow a syscall batching through
an fd, which can be executed when fd closes.

This project doesn't use any fd, the syscalls
are submitted via the batch() syscall. This
makes it a bit difficult to specify an exact
moment of execution, which, in case of an
fd, could be specified with fsync() and close().

So my question is: how difficult would it be
to provide the fd-specific interface for batching?
Does this look like a worthy idea? I think
it will allow a more precise control on the whole
process: you can add new syscalls to a queue
with a mere write(), you can start the execution
with fsync(), you can monitor the completion
via select(), and all that. I think the trend of an
fd-alike interfaces is uprising (signalfd, timerfd
and so on), so what do you think of all that?

**Comment by c-blake** [comment]
From a difficulty point of view, it does not seem that hard, but it'd
likely be a whole new API.  It may help to note one big `copy_from_user`
is much faster than many small ones.

From a worthiness point of view, I see little practical difference between:
  A) "building up an array to do a buffered `write` followed by `fsync`"
and
  B) "building up an array to then do a `batch`".

The trend toward fd-alike interfaces that you mention is more about a
unified idea of notification channels / events via `select` & friends
"joining" the "return" ctl flow from >=2 sources.  Callers still wait
for `select` to return.  Also, io_uring is arguably *not* in that trend,
yet also trendy. :)

You also mention `close` for A), but that also seems almost the same.
The name is just more suggestive of release/completion.  For example,
if you integration with C++ RAII activation via destructors then you
can just put the buffering up in the main logic and `batch` in the
destructor.  So, it might be "more natural because of the name" to think
of putting a `close` in resource clean up, but operationally it's not so
different.

Also, any given `batch` might well have a `close` within it, as with
https://github.com/c-blake/batch/blob/master/examples/total.c .  That
program also makes good use of the early exit & `wdcpy` minimal logic
features of `sys_batch` to form higher level abstractions.  This value
comes from latent structure in the world where people "build up"
abstract operations out of lower level APIs (which, if batching ability
is assumed, might well be even lower level primitives!).

A `batch` might also have a `select` or `poll` inside, although the
"word copy" operator might not be enough to make an easy "poll + read
ready into prepared buffers" combo syscall.  The logic ability itself
expands `sys_batch` use a lot while avoiding Halting Problem / proof
carrying code complexity (since back jumps are impossible, so are
loops, infinite or not), but I doubt it can easily extend to async.

All this said, if you do write something, I am happy to link to it from
here or vice-versa for others accidentally finding either bit of work.

**Reply by stsp** [reply]
> You also mention close for A), but that also seems almost the same.

This however would solve the "cleanup"
problem, because close() is called implicitly
even if the process is killed with SIGKILL.
There is no other way of running any cleanup
code after SIGKILL.

> The trend toward fd-alike interfaces that you mention is more about a
> unified idea of notification channels / events via select & friends

Exactly, and the one can add the batchfd
to his select() in order to be notified about
the batch completion.

Also with batchfd() an extra flags can be used.
For example the O_SYNC flag may mean
to initiate the batch as soon as possible,
rather than to wait for fsync() or close(),
which would make it exactly similar to the
current batch().

So your argument is that the execution
control is not needed because I can as
well call batch() later. This however doesn't
count the "cleanup" case, and also the ability
to get batch completion via select() looks
beneficial. No other advantages implied,
just these 2.

**Reply by c-blake** [reply]
When the kernel "closes" all fds at process destruction, it can release
resources *below* the syscall interface (e.g. at the VFS layer in Linux).

Also, `select` is *already* a batch interface, but batching for the
purpose of control flow "join" which as mentioned interferes with the
sync logic part of `sys_batch` (as does your O_ASYNC idea) which
adds value other ways.

So, I am just not seeing a good value trade-off here, but I don't want to
discourage you from your own experimentation.  Your batchfd may work
as an *additional* API, but I don't think it can really be a *replacement*
API for `sys_batch`.  batchfd seems less not more convenient/precise.

**Reply by stsp** [reply]
> When the kernel "closes" all fds at process destruction, it can release
> resources below the syscall interface (e.g. at the VFS layer in Linux).

Well, usually every kernel subsystems
has a `.release` member, so it should be
possible to make it process the batch at
a termination time, just as if the close()
was explicitly called.

> Also, select is already a batch interface

With that I am a bit puzzled because
earlier you said:
```
A batch might also have a select or poll inside
```
And I am not talking about the "inside".
Is there an example that waits for the
batch completion via select?

**Reply by c-blake** [reply]
I don't have such an example, but I think if you look through my
`examples/` it may become more clear.  It's really just `close` or
`select`, being syscalls themselves, can be in a `syscall_t array[]`.

What I meant by `select` already being a batch interface is that it
operates on a big batch of fds all at once - only to join control
flow against some condition, as opposed to `sys_batch` which is
about more arbitrary requests to the kernel.

In terms of `.release`, I think we may be not communicating well.
I mean "system call" as part of the kernel-user interface, not a
more generic "calls within the kernel/a 'system'" sense.  Really, we
may have an [XY Problem](https://en.wikipedia.org/wiki/XY_problem)
on the overall discussion.  So, we might progress better if you said more
about your broader context.

**Reply by stsp** [reply]
> It's really just close or
> select, being syscalls themselves, can be in a syscall_t array[].

But I don't need that (in this discussion
at least). I am talking about this:
```
int fd = batchfd(flags);
FD_SET(fd, &my_set);
/* now I can use select() to get a completion event from my batches */
```
This has nothing to do with select() being
a part of the batch.

> In terms of .release, I think we may be not communicating well.
> I mean "system call" as part of the kernel-user interface, not a
> more generic "calls within the kernel/a 'system'" sense.

Same problem: I am talking about the close()
on an fd returned by batchfd(), not about close()
being a part of the batch.
```
int fd = batchfd(flags);
write(fd, my_batch, sizeof(my_batch));
close(fd);
/* at that point batch definitely processed and completed */
```

Now this close may either be an explicit syscall,
or a resource clean-up done by the kernel. Its
the same thing, they both would call a .release
handler of a batchfd subsystem.

**Comment by c-blake** [comment]
I still do not know your broader problem context, but it is starting
to seem like what you want is an async system call / kernel service
request interface.  Any such must incidentally enable batching since
async mostly implies >1 outstanding request.  So, you should probably
learn about https://en.wikipedia.org/wiki/Io_uring (mentioned already).

I doubt that can address your resource release on process termination,
but there may be security implications in delegating any of that to
user-programs.  If none is delegated then it must already all be on
the kernel side of the interface and so both "already batched" and
also already handled.  For example, I might be wrong on this, but if
a `SIGKILL` schedules a process for teardown *during* a `sys_batch`
then I believe all the usual clean-up should still happen - whatever is
appropriate given however far the loop doing `sys_batch` calls got.

**Reply by stsp** [reply]
> I doubt that can address your resource release on process termination,

Yes, this doesn't. :)

> but there may be security implications in delegating any of that to
> user-programs.

But what delegation you mean?
I just want to batch a few `unlink()`
syscalls for a process termination.
With `batch()` they are executed
immediately, and with `batchfd()`
I can defer that to `close()`, which
happens on a process termination.
Why security implications?

**Comment by c-blake** [comment]
Delegation of resource creation for the purpose of later resource release.
These are a logical pair and proper clean-up is part of "security" to me,
as in not letting rogue processes run systems out of resources.

Now that you describe a little more what you want, the issue seems to
come down more to scheduling work on behalf of a process at its exit,
like standard C [`atexit`](https://en.wikipedia.org/wiki/C_process_control)
handlers, but at the kernel level instead of the libc level.  At that time,
the process itself may no longer be viable.  Consider `SIGSEGV` where any
memory reference the process makes might be invalid, including some
`syscall_t array[]` buffer which may be trash.

I haven't looked into process tear down in detail, but my best idea for you
is to try to "hook into" the kernel auto-close on process tear down to make
close on an fd connected to a new virtual FS (possibly via Linux FUSE)
do the desired bundle of operations (even upon a SIGKILL of the proc
with the open fd).  With this idea, as long as the kernel FS module and/or
user-space FS server remain viable, they can do the desired clean-up for
this new clean-up service.

Now, while you might use an API very similar to `sys_batch` to register
that work, it may be even nicer to register a `.so` with some C linkage
entry point / something.. or even a file or directory name to read at
process exit time that the process can update as it works.  Not sure - the
interfacing problem is just a less constrained one.. I am brainstorming on
your behalf. :-)  But I do now see the clean-up relevance & relation to
`sys_batch`.

{ This also touches on another thing `sys_batch` does *not* provide which
indeed would be nice -- namely "atomicity", like the atomicity of
`sys_rename` vs the `link/unlink` pair or of `readv` & `writev`.  Some
kind of generic new locking syscalls would be needed, I think, or some
agreement to use sysvipc things, and such would, in turn, probably open up
possibilities of deadlock.  Right now, I think these things & non-deadlock
are ensured "deep down" in kernel call stacks. }

**Reply by stsp** [reply]
> Consider SIGSEGV where any
> memory reference the process makes might be invalid, including some
> syscall_t array[] buffer which may be trash.

But the array was submitted before
SIGSEGV, so I don't think trashing
an array is a problem.
I suppose batched syscalls still take
their args from the user buffer, but eg
in case of `unlink()` the argument is
`const char *` and is usually pointing
the the read-only .data section that
can't be trashed.
Also if the syscall (even w/o batching)
hits the unmapped page, it just returns
EFAULT.

So really the risk looks very minor to
me, and also you can get the SIGSEGV
even without batchfd(), just when the
batch() is processing. So I think the
SIGSEGV problem is unspecific to my
use-case. Memory corruption can happen
in any case, SIGSEGV may happen in
any case, and batched syscalls may try
to access the corrupted data even in your
current impl.

**Reply by c-blake** [reply]
> But the array was submitted before

Well, sys_batch copies the array to work on it right away.
Your "deferred until process tear down" work would have to
allocate the memory for a maybe long time, and since the
maybe buggy process gets control again might want to
"copy deeper", like copy any "indirect data" chasing the
char*'s.  (These are just considerations, not deal-breakers.)

> const char * ... pointing the the read-only .data section

The `const` only means `unlink` "need not" / should not change it.
It could easily be a string calculated from run-time data.  String
literals may be in .rodata.  If you only `unlink` string literals,
well, that is pretty specific to your own use cases. :-)  Many may
take directories from config files or add filename extensions, etc.

**Reply by stsp** [reply]
> "copy deeper", like copy any "indirect data" chasing the

I wouldn't do that.
Yes, there is a risk to hit the corruption,
but the same is true even now. You can
get SIGSEGV (or a corruption without
SIGSEGV) while batch is executing.
So I don't actually see why that problem
is specific to my case.

**Reply by stsp** [reply]
Or wait, are the batches completely
synchronous now?

**Reply by stsp** [reply]
Sorry, your README doesn't say whether
the batches are synchronous or not, so I was
assuming they are async.
I guess that was a wrong assumption?

**Reply by c-blake** [reply]
Yes - right now sys_batch is completely synchronous.  Once
called it does not return to the user program until the end of
the batch.  I was speculating above you might like io_uring.
That may require "amendments", but that might also provide
a good "hook" into process tear down work registration.  It
might be a simpler addition there than re-doing sys_batch.

**Reply by c-blake** [reply]
(Although if you pursue that FUSE idea then there may be
no amendment to Linux at all required, but you might need
an extra file server process.)

README-wise, I feel like async / anything out of the usual
call-return model is special and requires the extra work, as
well as describing that work.  So, I did not hit the reader over
the head with "what sys_batch *wasn't*". :-)

**Reply by stsp** [reply]
> might be a simpler addition there than re-doing sys_batch.

Yes, sorry for bothering you with this.
Being synchronous, batch() is completely
out of any considerations for the batchfd()
API. They can co-exist together if someone
implements the later.

FUSE module may work, although its
still far from trivial. Maybe it should be
some virtual device implemented in
fuse and accessed with ioctl()s, not sure...

**Reply by stsp** [reply]
https://lwn.net/Articles/261473/
So it was implemented at one point,
but not sure of its destiny today. :(

**Comment by c-blake** [comment]
I don't recall ever hearing about syslets making it into mainline.
My first attempt at batch in Linux back circa 2000 used that
same `indirect` call mechanism syslets use.  When I did this
version, it seemed that call was gone/also had an ill-fated destiny.
So, I just rebuilt what I needed to make it work.
